fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+			}
+		})
+
+		It("v3 node resource annotations should automatically heal", func() {
+			for _, felix := range felixes {
+				// Get the original public-key.
+				node, err := client.Nodes().Get(context.Background(), felix.Hostname, options.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				wgPubKeyOrig := node.Status.WireguardPublicKey
+
+				// overwrite public-key by fake but valid wireguard key.
+				node.Status.WireguardPublicKey = fakeWireguardPubKey
+				_, err = client.Nodes().Update(context.Background(), node, options.SetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				Eventually(func() string {
+					node, err := client.Nodes().Get(context.Background(), felix.Hostname, options.GetOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					return node.Status.WireguardPublicKey
+				}, "5s", "100ms").Should(Equal(wgPubKeyOrig))
+			}
+		})
+
+		It("workload to workload traffic should be allowed and encrypted", func() {
+			for _, felix := range felixes {
+				// Start tcpdump on each "host".
+				tcpdump := felix.AttachTCPDump("eth0")
+				inTunnelPacketsPattern := fmt.Sprintf("IP %s\\.51820 > \\d+\\.\\d+\\.\\d+\\.\\d+\\.51820: UDP", felix.IP)
+				tcpdump.AddMatcher("numInTunnelPackets", regexp.MustCompile(inTunnelPacketsPattern))
+				outTunnelPacketsPattern := fmt.Sprintf("IP \\d+\\.\\d+\\.\\d+\\.\\d+\\.51820 > %s\\.51820: UDP", felix.IP)
+				tcpdump.AddMatcher("numOutTunnelPackets", regexp.MustCompile(outTunnelPacketsPattern))
+				tcpdump.Start()
+
+				tcpdumps = append(tcpdumps, tcpdump)
+			}
+
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+
+			By("verifying tunnelled packet count")
+			for i := range felixes {
+				Eventually(tcpdumpMatchCount(i, "numInTunnelPackets")).Should(BeNumerically(">", 0))
+				Eventually(tcpdumpMatchCount(i, "numOutTunnelPackets")).Should(BeNumerically(">", 0))
+			}
+
+			By("verifying peer tunnel transmit details are complimentary")
+			var sent, rcvd [nodeCount]string
+			for i, felix := range felixes {
+				out, _ := felix.ExecOutput("wg")
+				matches := xferRegExp.FindStringSubmatch(out)
+				Expect(matches).NotTo(BeNil())
+				rcvd[i] = matches[1]
+				sent[i] = matches[2]
+			}
+			Expect(rcvd[0]).To(Equal(sent[1]))
+			Expect(rcvd[1]).To(Equal(sent[0]))
+		})
+	})
+
+	Context("with Wireguard disabled", func() {
+		BeforeEach(func() {
+			felixPID := felixes[0].GetFelixPID()
+
+			enableDisableWireguard(client, false)
+
+			// Wait for felix to restart.
+			Eventually(felixes[0].GetFelixPID, "5s", "100ms").ShouldNot(Equal(felixPID))
+		})
+
+		It("the Wireguard device shouldn't exist", func() {
+			for _, felix := range felixes {
+				Eventually(func() string {
+					out, _ := felix.ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+		})
+
+		It("the Wireguard rule-entry/route-table shouldn't exist", func() {
+			for i, felix := range felixes {
```

Comment:
I mention this below, but I think rule and route checks are different things - much cleaner if you split them out.